### DISCUSSION TITLE
[mini.deps] Apply Folke's Tokyonight Dev Environment

### DISCUSSION THREAD

**Discussion by abeldekat** [nvim-mini/mini.nvim#1012] [Show and tell]
Before:

![1720081830](https://github.com/echasnovski/mini.nvim/assets/58370433/c61bd63e-b9ae-4876-b865-13c9c0554b74)

After:

![1720081671](https://github.com/echasnovski/mini.nvim/assets/58370433/0819f8a2-d616-469a-9166-3c433bf58cbb)

As the [initiator](https://github.com/LazyVim/LazyVim/pull/2115) of `.lazy.lua` I was curious to see if the [Tokyonight Dev Environment](https://www.reddit.com/r/neovim/comments/1dtk0iq/tokyonightnvim_v4x_rewrite/) could also be applied using `mini.deps` as the package manager.

This is especially interesting when other `colorscheme` creators also [apply](https://www.reddit.com/r/neovim/comments/1duuo6r/folke_has_set_the_bar_for_neovim_dx_yet_again/) this approach.

The code:

```lua
local hipatterns = require("mini.hipatterns")

-- Prepend a space, append a colon:
local function to_pattern(word) return " " .. word .. ":" end

-- Lots of hightlighers....
local function get_highlighters()
  local highlighters = {
    FIX = { pattern = to_pattern("FIX"), group = "MiniHipatternsFixme" },
  }
  return highlighters
end

--[[
Enable project-specific plugin specs.

File .lazy.lua:
  is read when present in the current working directory
  should return a plugin spec
  has to be manually trusted for each instance of the file

See:
  :h 'exrc'
  :h :trust
--]]
---@return function?
local function get_opt_function_from_lazy_lua()
  local filepath = vim.fn.fnamemodify(".lazy.lua", ":p")
  local file = vim.secure.read(filepath)
  if not file then return end

  local package = vim.fn.fnamemodify(filepath, ":p:h:t")
  vim.cmd.packadd(package) -- name of dir holding the colorscheme to packadd

  local lazyspec = loadstring(file)()
  if not lazyspec then return end

  local hi_spec = lazyspec[1]
  if not hi_spec or type(hi_spec[1]) ~= "string" or not string.find(hi_spec[1], "hipatterns") then return end

  local hi_spec_callable = hi_spec.opts
  if not hi_spec_callable or type(hi_spec_callable) ~= "function" then return end

  return hi_spec_callable
end

local highlighters = get_highlighters()
local lazy_opt_function = get_opt_function_from_lazy_lua()
local opts = { highlighters = highlighters }
hipatterns.setup(lazy_opt_function and lazy_opt_function(nil, opts) or opts)

```

Open a file inside the directory of `tokyonight.nvim` and the patterns apply exactly as when using `lazy.nvim`. An initial `:w` might be necessary.

**Comment by pkazmier** [comment]
While I don't plan on visiting tokyonight or cyberdreams repos anytime soon, I still added this to my config 😄

Made a small change and replaced these two lines:
```lua
  local hi_spec = lazyspec[1]
  if not hi_spec or type(hi_spec[1]) ~= "string" or not string.find(hi_spec[1], "hipatterns") then return end
```
With these lines just in case someone has other plug-in specs in their `.lazy.lua` file:
```lua
  local hi_spec = vim.tbl_filter(function(spec)
    return type(spec[1]) == "string" and spec[1]:find("hipatterns") ~= nil
  end, lazyspec)[1]
  if not hi_spec then return end
```

**Reply by abeldekat** [reply]
Very nice! Thanks for the improvement.

**Comment by echasnovski** [comment]
Thanks for sharing!

I am thinking about an idea of adding something like `MiniExtra.gen_highlighter.hl_groups` to possibly make this easier. So that it highlight whole word in buffer if it is equal to *some* present highlight group. Will need an investigation on performant implementation (some kind of caching for present highlight groups, at least). We'll see.

**Reply by abeldekat** [reply]
That would be great!

**Reply by abeldekat** [reply]
Just a thought: Would this extra have the same intent as the code Folke wrote in `.lazy.lua`? If that's the case, and the extra is used in a local `.lazy.lua` in the colorscheme, performance is perhaps not that critical.

Would an "extra", reading the local `.lazy.lua` be interesting? The `mini.deps` user could then select from the output of the extra when doing `require("someplugin").setup("opts_with_extra_opts")`

**Reply by echasnovski** [reply]
> Just a thought: Would this extra have the same intent as the code Folke wrote in `.lazy.lua`? If that's the case, and the extra is used in a local `.lazy.lua` in the colorscheme, performance is perhaps not that critical.
> 
> Would an "extra", reading the local `.lazy.lua` be interesting? The `mini.deps` user could then select from the output of the extra when doing `require("someplugin").setup("opts_with_extra_opts")`

The intent is to make it easier to have manually toggled highlighter which highlights highlight group names inside a buffer. Reading '.lazy.lua' is probably too much here.
